packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -169,14 +222,70 @@ class _SegmentedControlState<T> extends State<SegmentedControl<T>> {
   }
 
   void _onTap(T currentKey) {
-    if (currentKey != widget.groupValue) {
+    if (currentKey != widget.groupValue && currentKey == _pressedKey) {
       widget.onValueChanged(currentKey);
+      _pressedKey = null;
+    }
+  }
+
+  Color getTextColor(int index, T currentKey) {
+    if (_selectionControllers[index].isAnimating) {
+      return textColorTween.evaluate(_selectionControllers[index]);
+    } else if (widget.groupValue == currentKey) {
+      return CupertinoColors.white;
+    }
+    return CupertinoColors.activeBlue;
+  }
+
+  Color getBackgroundColor(int index, T currentKey) {
+    if (_selectionControllers[index].isAnimating) {
+      return _childTweens[index].evaluate(_selectionControllers[index]);
+    } else if (widget.groupValue == currentKey) {
+      return CupertinoColors.activeBlue;
+    } else if (_pressedKey == currentKey) {
+      return _kPressedBackground;
+    }
+    return CupertinoColors.white;
+  }
+
+  void updateAnimationControllers() {
+    while (_selectionControllers.length != widget.children.length) {
+      if (_selectionControllers.length > widget.children.length) {
+        _selectionControllers.removeLast();
+        _childTweens.removeLast();
+      } else {
+        _selectionControllers.add(createAnimationController());
+      }
+    }
+  }
+
+  @override
+  void didUpdateWidget(SegmentedControl<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+
+    if (oldWidget.children.length != widget.children.length) {
+      updateAnimationControllers();
+    }
+
+    if (oldWidget.groupValue != widget.groupValue) {
+      int index = 0;
+      for (T key in widget.children.keys) {
+        if (widget.groupValue == key) {
+          _childTweens.insert(index, forwardBackgroundColorTween);
+          _selectionControllers[index].forward();
+        } else {
+          _childTweens.insert(index, reverseBackgroundColorTween);
+          _selectionControllers[index].reverse();
+        }
+        index += 1;
```

Comment:
isn't this going to end up inserting a whole bunch of tweens each time the value changes?